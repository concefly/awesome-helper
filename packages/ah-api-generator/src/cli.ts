#!/usr/bin/env node

import yargs from 'yargs';
import { generateAPIService } from './index';

const args = yargs
  .option('input', { alias: 'i', describe: 'apiDoc', type: 'string', required: true })
  .option('headers', { describe: 'fetch headers', type: 'array' })
  .option('operation-id-splitter', { describe: 'operationId 分隔符，默认', type: 'string' })
  .option('template', { alias: 't', describe: '模板', type: 'string' })
  .option('banner', { alias: 'b', describe: 'banner', type: 'string' })
  .option('dump', { alias: 'd', describe: '导出', type: 'string' }).argv;

generateAPIService({
  input: args.input.match(/^(http|https)\:\/\//)
    ? {
        type: 'remote',
        url: args.input,
        headers: args.headers
          ? args.headers.reduce((re, cu) => {
              const [k, v] = (cu + '').split(':');
              return { ...re, [k]: v };
            }, {} as any)
          : undefined,
      }
    : { type: 'local', filename: args.input },
  operationIdSplitter: args['operation-id-splitter'],
  template: args.template,
  banner: args.banner,
  dump: args.dump,
})
  .then(content => {
    if (!args.dump) console.log(content);
  })
  .catch(err => {
    console.error(err);
    process.exit(1);
  });
