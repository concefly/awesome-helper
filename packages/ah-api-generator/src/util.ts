import { OpenAPIV3 as API } from 'openapi-types';
import _ from 'lodash';
import { Schema } from 'jsonschema';

import { inspect } from 'util';

export const log = (...args: any[]) => {
  console.log(...args.map(arg => inspect(arg, { showHidden: false, depth: null })));
};

export function parseRequestSchema(
  data:
    | API.PathItemObject['get']
    | API.PathItemObject['put']
    | API.PathItemObject['post']
    | API.PathItemObject['delete']
): Schema | undefined {
  if (!data) return;

  const schema: Schema = {
    type: 'object',
    properties: {},
  };

  if (data.parameters) {
    (data.parameters as API.ParameterObject[]).forEach(p => {
      Object.assign(schema.properties!, {
        [p.name]: p.schema,
      });
    });
  }

  if (data.requestBody) {
    const requestBodyContentJson = (data.requestBody as API.RequestBodyObject).content[
      'application/json'
    ];

    if (typeof (requestBodyContentJson?.schema as any)?.type !== 'undefined') {
      const jSchema = requestBodyContentJson?.schema as Schema;
      _.merge(schema, jSchema);
    }
  }

  // 没有收集到，返回空
  if (_.isEmpty(schema.properties)) return;

  return schema;
}

export function parseResponseSchema(
  data:
    | API.PathItemObject['get']
    | API.PathItemObject['put']
    | API.PathItemObject['post']
    | API.PathItemObject['delete']
): Schema | undefined {
  const response = data?.responses?.['200'] as API.ResponseObject;
  const jsonResponse = response.content?.['application/json'];

  if (jsonResponse) {
    // 尝试取 schema
    if (jsonResponse.schema) return jsonResponse.schema as Schema;

    const example =
      jsonResponse.example || (jsonResponse.examples && Object.values(jsonResponse.examples).pop());

    // 尝试转换 example
    if (example) return data2Schema(example);
  }
}

export type IRequestMeta = {
  query?: {
    schema: Schema;
    tsTypeLiteral: string;
  };
  response?: {
    schema: Schema;
    tsTypeLiteral: string;
  };
  pathName: string;
  method: string;
  description?: string;
};

export interface IServiceMap {
  [serviceName: string]: {
    [requestName: string]: IRequestMeta;
  };
}

export function renderService(apiDoc: API.Document, opt: { operationIdSplitter?: string } = {}) {
  const serviceMap: IServiceMap = {};
  const { operationIdSplitter = '.' } = opt;

  Object.entries(apiDoc.paths).forEach(([pathName, pi]) => {
    if (!pi) return;

    Object.entries({
      GET: pi.get,
      PUT: pi.put,
      POST: pi.post,
      DELETE: pi.delete,
    }).forEach(([method, commonPiData]) => {
      if (!commonPiData) return;

      let [serviceName, requestName] = (
        commonPiData.operationId || `default${operationIdSplitter}${_.camelCase(pathName)}`
      ).split(operationIdSplitter);

      serviceName = _.capitalize(serviceName);

      const requestSchema = parseRequestSchema(commonPiData);
      const respSchema = parseResponseSchema(commonPiData);

      const requestMeta: IRequestMeta = {
        query: requestSchema
          ? { schema: requestSchema, tsTypeLiteral: schema2TsTypeLiteral(requestSchema) }
          : undefined,
        response: respSchema
          ? { schema: respSchema, tsTypeLiteral: schema2TsTypeLiteral(respSchema) }
          : undefined,
        method,
        pathName,
        description: commonPiData.description,
      };

      //
      _.set(serviceMap, [serviceName, requestName], requestMeta);
    });
  });

  return serviceMap;
}

export function schema2TsTypeLiteral(s?: Schema): string {
  if (!s) return 'never';

  if (s.type === 'string') return 'string';
  if (s.type === 'integer') return 'number';

  if (s.type === 'array') {
    return `Array<${schema2TsTypeLiteral((s.items as any) || '{}')}>`;
  }

  if (s.type === 'object') {
    {
      if (s.properties) {
        return [
          '{',
          Object.entries(s.properties).map(([pn, pv]) => {
            const isRequired =
              typeof s.required === 'boolean'
                ? s.required
                : Array.isArray(s.required)
                ? s.required.includes(pn)
                : false;

            const tsType = schema2TsTypeLiteral(pv);

            return `${pn}${isRequired ? '' : '?'}: ${tsType}`;
          }),
          '}',
        ].join(' ');
      }

      return 'object';
    }
  }

  return 'any';
}

/** 从数据结构猜测 schema */
export function data2Schema(data: any): Schema {
  if (_.isString(data)) return { type: 'string' };
  if (_.isInteger(data)) return { type: 'integer' };
  if (_.isNumber(data)) return { type: 'number' };

  if (_.isBoolean(data)) return { type: 'boolean' };

  if (_.isArray(data)) {
    return {
      type: 'array',
      items: data2Schema(data[0]),
    };
  }

  if (_.isPlainObject(data)) {
    return {
      type: 'object',
      properties: _.mapValues(data, data2Schema),
      required: Object.keys(data),
    };
  }

  return {};
}
