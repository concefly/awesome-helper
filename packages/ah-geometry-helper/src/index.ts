export * from './IDirType';
export * from './Line';
export * from './LineSegment';
export * from './LineSet';
export * from './Point';
export * from './Rect';
