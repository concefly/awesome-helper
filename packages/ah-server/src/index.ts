import 'koa';
import { DefaultContext, DefaultState, ParameterizedContext } from 'koa';

export interface IApplication {}
export interface IService {}

export interface IContext extends ParameterizedContext<DefaultState, DefaultContext> {}
export interface IConfig {
  LOCAL_PORT: number;
}

// fix koa plugin type
declare module 'koa' {
  interface Request {
    // koaBody type
    body?: any;
  }
}

export * from './App';
export * from './Config';
export * from './Logger';
export * from './Controller';
export * from './Service';
export * from './util';
export * from './Scheduler';
export * from './error';
