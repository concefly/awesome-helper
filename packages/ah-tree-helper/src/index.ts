export * from './TreeHelper';
export * from './Dictionary';
export * from './ITreeNode';
