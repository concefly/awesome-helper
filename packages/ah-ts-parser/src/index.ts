export * from 'ts-morph';
export * from './Ts2Schema';
export * from './util';
