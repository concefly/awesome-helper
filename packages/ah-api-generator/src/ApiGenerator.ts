import _ from 'lodash';
import { OpenAPIV3 as API } from 'openapi-types';
import { IRequestMeta, IServiceMap, renderService } from './util';

interface IConfig {
  renderRequestMethod(ctx: { requestName: string; meta: IRequestMeta }): string;

  renderServiceCls(ctx: {
    StartCaseServiceName: string;
    camelCaseServiceName: string;
    baseServiceName: string;
    requestMethodLiterals: string[];
  }): string;

  renderFileContent(ctx: { serviceClsLiterals: string[]; serviceMap: IServiceMap }): string;
}

/** API 生成器 */
export class ApiGenerator {
  private config: IConfig = {
    renderRequestMethod: ctx => {
      const cDesc = `/** ${ctx.meta.description || ctx.requestName} */`;
      const crCallArgs = `pathName: '${ctx.meta.pathName}', method: '${ctx.meta.method}'`;

      if (!ctx.meta.query) {
        return `
${cDesc}
${ctx.requestName} = async () => { return this.request({ ${crCallArgs} }); }`;
      }

      const schemaLiteral = JSON.stringify(ctx.meta.query.schema);

      return `
${cDesc}
${ctx.requestName} = async (query: ${ctx.meta.query.tsTypeLiteral}) => {
  return this.request({ ${crCallArgs}, query, schema: ${schemaLiteral} });
}`;
    },

    renderServiceCls: ctx => `
export class ${ctx.StartCaseServiceName}Service extends ${ctx.baseServiceName} {
  ${[ctx.requestMethodLiterals.join('\n')]}      
}`,

    renderFileContent: ctx => ctx.serviceClsLiterals.join('\n\n'),
  };

  constructor(readonly apiDoc: API.Document, _config: Partial<IConfig> = {}) {
    Object.assign(this.config, _config);
  }

  private serviceMap = renderService(this.apiDoc);

  /** 导出成字符串 */
  emitToString(opt: { baseServiceName?: string } = {}) {
    const { baseServiceName = 'BaseService' } = opt;

    const serviceClsLiterals = Object.entries(this.serviceMap).map(([serviceName, serviceData]) => {
      const requestMethodLiterals = Object.entries(serviceData).map(([requestName, reqMeta]) => {
        // 生成请求函数
        return this.config.renderRequestMethod({
          requestName,
          meta: reqMeta,
        });
      });

      // 生成 service
      return this.config.renderServiceCls({
        StartCaseServiceName: _.startCase(serviceName).replace(/\s+/g, ''),
        camelCaseServiceName: _.camelCase(serviceName),
        baseServiceName,
        requestMethodLiterals,
      });
    });

    return this.config.renderFileContent({ serviceClsLiterals, serviceMap: this.serviceMap });
  }

  getServiceMap() {
    return _.cloneDeep(this.serviceMap);
  }
}
