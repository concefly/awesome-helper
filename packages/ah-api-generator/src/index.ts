import path from 'path';
import fs from 'fs';
import { curl } from 'urllib';
import { renderService } from './util';
import ejs from 'ejs';
import { OpenAPIV3 as API } from 'openapi-types';
import prettier from 'prettier';
import RefParser from '@apidevtools/json-schema-ref-parser';

export interface IConfig {
  input?: { type: 'local'; filename: string } | { type: 'remote'; url: string; headers?: any };
  operationIdSplitter?: string;
  template?: string;
  banner?: string;
  dump?: string;
}

export const loadApiDoc = async (input: IConfig['input']): Promise<API.Document> => {
  if (!input) throw new Error('input 不存');

  let result =
    input.type === 'local'
      ? JSON.parse(fs.readFileSync(input.filename, { encoding: 'utf-8' }))
      : input.type === 'remote'
      ? await curl<any>(input.url, { headers: input.headers, dataType: 'json' }).then(r => r.data)
      : null;

  if (!result) throw new Error(`inputSource 不存在: ${JSON.stringify(input)}`);

  // 解 $ref
  result = await RefParser.dereference(result);

  return result;
};

export const generateAPIService = async (cfg: IConfig) => {
  const {
    input = { type: 'local', filename: 'api.json' },
    template = path.resolve(__dirname, '..', 'template.ejs'),
    dump,
    banner,
  } = cfg;

  const apiDoc = await loadApiDoc(input);
  const serviceMap = renderService(apiDoc, { operationIdSplitter: cfg.operationIdSplitter });

  // render template
  let content = ejs.render(fs.readFileSync(template, { encoding: 'utf-8' }), { serviceMap });
  if (banner) content = banner + '\n' + content;

  // pretty
  const prettierOpt = await prettier.resolveConfig(process.cwd());
  content = prettier.format(content, {
    parser: 'typescript',
    trailingComma: 'es5',
    tabWidth: 2,
    semi: true,
    singleQuote: true,
    jsxSingleQuote: true,
    printWidth: 100,
    bracketSpacing: true,
    arrowParens: 'avoid',
    ...prettierOpt,
  });

  // dump
  if (dump) fs.writeFileSync(dump, content, { encoding: 'utf-8' });

  return content;
};
