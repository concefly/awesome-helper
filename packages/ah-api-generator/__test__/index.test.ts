import { generateAPIService } from '../src';

describe('generateAPIService', () => {
  [{ name: 'api' }, { name: 'api2', operationIdSplitter: '_' }].forEach(
    ({ name, operationIdSplitter }) => {
      it(name, async () => {
        const tsContent = await generateAPIService({
          input: { type: 'local', filename: `${__dirname}/${name}.json` },
          banner: '//banner',
          operationIdSplitter,
        });
        expect(tsContent).toMatchSnapshot(name);
      });
    }
  );
});
