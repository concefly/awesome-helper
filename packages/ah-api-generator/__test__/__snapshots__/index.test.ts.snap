// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateAPIService api: api 1`] = `
"//banner
import { BaseService } from './BaseService';

export type IPageService_getList_Query = { name?: string; title?: string; summary?: string };
export type IPageService_getList_Resp = {
  pageSize: number;
  pageNum: number;
  total: number;
  list: Array<{
    id: number;
    created_at: string;
    updated_at: string;
    level: number;
    a: string;
    b: any;
    c: any;
    e: any;
  }>;
};

export type IPageService_createOne_Query = {
  title: string;
  summary: string;
  content: string;
  meta: object;
};
export type IPageService_createOne_Resp = any;

export type IPageService_updateOne_Query = {
  name: string;
  title?: string;
  summary?: string;
  content?: string;
  meta?: object;
};
export type IPageService_updateOne_Resp = any;

export type IPageService_removeOne_Query = { name?: string };
export type IPageService_removeOne_Resp = any;

export type IReserveService_getActivityList_Query = { id?: number };
export type IReserveService_getActivityList_Resp = any;

export type IReserveService_createActivity_Query = {
  title: string;
  detail?: string;
  meta?: { richDetailMeta?: object };
};
export type IReserveService_createActivity_Resp = any;

export type IReserveService_updateActivity_Query = {
  id: number;
  title?: string;
  detail?: string;
  meta?: { richDetailMeta?: object };
};
export type IReserveService_updateActivity_Resp = any;

export type IReserveService_removeActivity_Query = { id?: number };
export type IReserveService_removeActivity_Resp = any;

export type IReserveService_getCalenderList_Query = {
  id?: number;
  activity_id?: number;
  startTime?: string;
  endTime?: string;
};
export type IReserveService_getCalenderList_Resp = any;

export type IReserveService_createCalender_Query = {
  activity_id: number;
  list: Array<{
    start_time: string;
    end_time: string;
    location?: string;
    detail?: string;
    status?: string;
    meta?: { richDetailMeta?: object };
  }>;
};
export type IReserveService_createCalender_Resp = any;

export type IReserveService_updateCalender_Query = {
  id: number;
  start_time?: string;
  end_time?: string;
  location?: string;
  detail?: string;
  status?: string;
  meta?: { richDetailMeta?: object };
};
export type IReserveService_updateCalender_Resp = any;

export type IReserveService_getUserRequestList_Query = { calendar_id?: number };
export type IReserveService_getUserRequestList_Resp = any;

export type IReserveService_confirmUserRequest_Query = { id: number; username: string };
export type IReserveService_confirmUserRequest_Resp = any;

export type ISystemService_getApiDoc_Query = any;
export type ISystemService_getApiDoc_Resp = any;

export type ISystemService_getList_Query = any;
export type ISystemService_getList_Resp = any;

export type ISystemService_updateOne_Query = any;
export type ISystemService_updateOne_Resp = any;

export type IUserService_getUserInfo_Query = any;
export type IUserService_getUserInfo_Resp = any;

export type IUserService_updateSession_Query = any;
export type IUserService_updateSession_Resp = any;

export type IUserService_getShiluRoles_Query = any;
export type IUserService_getShiluRoles_Resp = any;

export type IUserService_assignShiluRoles_Query = { roleId: number };
export type IUserService_assignShiluRoles_Resp = any;

export type IUserService_removeShiluRoles_Query = { roleId?: number };
export type IUserService_removeShiluRoles_Resp = any;

export type IUserService_getEnterCode_Query = { pavilionId?: number };
export type IUserService_getEnterCode_Resp = any;

export type IUserService_getLeaveInfo_Query = { pavilionId?: number };
export type IUserService_getLeaveInfo_Resp = any;

export type IUserService_addReserveRequest_Query = {
  calendar_id: number;
  type: string;
  meta?: { members?: Array<{ nick: string; gender: string; phone: string }> };
};
export type IUserService_addReserveRequest_Resp = any;

export type IUserService_cancelReserveRequest_Query = { id: number };
export type IUserService_cancelReserveRequest_Resp = any;

export class PageService extends BaseService {
  /**
   * PageService.getList
   *
   * - description:
   * - pathName: /page
   * - method: GET
   */
  public getList = async (query: IPageService_getList_Query) => {
    return this.request<IPageService_getList_Resp>(replaceUrlParams('/page', query), 'GET', query);
  };

  /**
   * PageService.createOne
   *
   * - description:
   * - pathName: /page
   * - method: POST
   */
  public createOne = async (query: IPageService_createOne_Query) => {
    return this.request<IPageService_createOne_Resp>(
      replaceUrlParams('/page', query),
      'POST',
      query
    );
  };

  /**
   * PageService.updateOne
   *
   * - description:
   * - pathName: /page/:name
   * - method: PUT
   */
  public updateOne = async (query: IPageService_updateOne_Query) => {
    return this.request<IPageService_updateOne_Resp>(
      replaceUrlParams('/page/:name', query),
      'PUT',
      query
    );
  };

  /**
   * PageService.removeOne
   *
   * - description:
   * - pathName: /page/:name
   * - method: DELETE
   */
  public removeOne = async (query: IPageService_removeOne_Query) => {
    return this.request<IPageService_removeOne_Resp>(
      replaceUrlParams('/page/:name', query),
      'DELETE',
      query
    );
  };
}

export class ReserveService extends BaseService {
  /**
   * ReserveService.getActivityList
   *
   * - description:
   * - pathName: /reserve_activity
   * - method: GET
   */
  public getActivityList = async (query: IReserveService_getActivityList_Query) => {
    return this.request<IReserveService_getActivityList_Resp>(
      replaceUrlParams('/reserve_activity', query),
      'GET',
      query
    );
  };

  /**
   * ReserveService.createActivity
   *
   * - description:
   * - pathName: /reserve_activity
   * - method: POST
   */
  public createActivity = async (query: IReserveService_createActivity_Query) => {
    return this.request<IReserveService_createActivity_Resp>(
      replaceUrlParams('/reserve_activity', query),
      'POST',
      query
    );
  };

  /**
   * ReserveService.updateActivity
   *
   * - description:
   * - pathName: /reserve_activity/:id
   * - method: PUT
   */
  public updateActivity = async (query: IReserveService_updateActivity_Query) => {
    return this.request<IReserveService_updateActivity_Resp>(
      replaceUrlParams('/reserve_activity/:id', query),
      'PUT',
      query
    );
  };

  /**
   * ReserveService.removeActivity
   *
   * - description:
   * - pathName: /reserve_activity/:id
   * - method: DELETE
   */
  public removeActivity = async (query: IReserveService_removeActivity_Query) => {
    return this.request<IReserveService_removeActivity_Resp>(
      replaceUrlParams('/reserve_activity/:id', query),
      'DELETE',
      query
    );
  };

  /**
   * ReserveService.getCalenderList
   *
   * - description:
   * - pathName: /reserve_calendar
   * - method: GET
   */
  public getCalenderList = async (query: IReserveService_getCalenderList_Query) => {
    return this.request<IReserveService_getCalenderList_Resp>(
      replaceUrlParams('/reserve_calendar', query),
      'GET',
      query
    );
  };

  /**
   * ReserveService.createCalender
   *
   * - description:
   * - pathName: /reserve_calendar
   * - method: POST
   */
  public createCalender = async (query: IReserveService_createCalender_Query) => {
    return this.request<IReserveService_createCalender_Resp>(
      replaceUrlParams('/reserve_calendar', query),
      'POST',
      query
    );
  };

  /**
   * ReserveService.updateCalender
   *
   * - description:
   * - pathName: /reserve_calendar/:id
   * - method: PUT
   */
  public updateCalender = async (query: IReserveService_updateCalender_Query) => {
    return this.request<IReserveService_updateCalender_Resp>(
      replaceUrlParams('/reserve_calendar/:id', query),
      'PUT',
      query
    );
  };

  /**
   * ReserveService.getUserRequestList
   *
   * - description:
   * - pathName: /reserve_request
   * - method: GET
   */
  public getUserRequestList = async (query: IReserveService_getUserRequestList_Query) => {
    return this.request<IReserveService_getUserRequestList_Resp>(
      replaceUrlParams('/reserve_request', query),
      'GET',
      query
    );
  };

  /**
   * ReserveService.confirmUserRequest
   *
   * - description:
   * - pathName: /reserve_request/:id/confirm
   * - method: PUT
   */
  public confirmUserRequest = async (query: IReserveService_confirmUserRequest_Query) => {
    return this.request<IReserveService_confirmUserRequest_Resp>(
      replaceUrlParams('/reserve_request/:id/confirm', query),
      'PUT',
      query
    );
  };
}

export class SystemService extends BaseService {
  /**
   * SystemService.getApiDoc
   *
   * - description:
   * - pathName: /system/apiDoc
   * - method: GET
   */
  public getApiDoc = async () => {
    return this.request<ISystemService_getApiDoc_Resp>(replaceUrlParams('/system/apiDoc'), 'GET');
  };

  /**
   * SystemService.getList
   *
   * - description:
   * - pathName: /system
   * - method: GET
   */
  public getList = async () => {
    return this.request<ISystemService_getList_Resp>(replaceUrlParams('/system'), 'GET');
  };

  /**
   * SystemService.updateOne
   *
   * - description:
   * - pathName: /system/:name
   * - method: PUT
   */
  public updateOne = async () => {
    return this.request<ISystemService_updateOne_Resp>(replaceUrlParams('/system/:name'), 'PUT');
  };
}

export class UserService extends BaseService {
  /**
   * UserService.getUserInfo
   *
   * - description:
   * - pathName: /user
   * - method: GET
   */
  public getUserInfo = async () => {
    return this.request<IUserService_getUserInfo_Resp>(replaceUrlParams('/user'), 'GET');
  };

  /**
   * UserService.updateSession
   *
   * - description:
   * - pathName: /user/updateSession
   * - method: GET
   */
  public updateSession = async () => {
    return this.request<IUserService_updateSession_Resp>(
      replaceUrlParams('/user/updateSession'),
      'GET'
    );
  };

  /**
   * UserService.getShiluRoles
   *
   * - description:
   * - pathName: /user/shiluRoles
   * - method: GET
   */
  public getShiluRoles = async () => {
    return this.request<IUserService_getShiluRoles_Resp>(
      replaceUrlParams('/user/shiluRoles'),
      'GET'
    );
  };

  /**
   * UserService.assignShiluRoles
   *
   * - description:
   * - pathName: /user/shiluRoles
   * - method: POST
   */
  public assignShiluRoles = async (query: IUserService_assignShiluRoles_Query) => {
    return this.request<IUserService_assignShiluRoles_Resp>(
      replaceUrlParams('/user/shiluRoles', query),
      'POST',
      query
    );
  };

  /**
   * UserService.removeShiluRoles
   *
   * - description:
   * - pathName: /user/shiluRoles
   * - method: DELETE
   */
  public removeShiluRoles = async (query: IUserService_removeShiluRoles_Query) => {
    return this.request<IUserService_removeShiluRoles_Resp>(
      replaceUrlParams('/user/shiluRoles', query),
      'DELETE',
      query
    );
  };

  /**
   * UserService.getEnterCode
   *
   * - description:
   * - pathName: /user/enterCode
   * - method: GET
   */
  public getEnterCode = async (query: IUserService_getEnterCode_Query) => {
    return this.request<IUserService_getEnterCode_Resp>(
      replaceUrlParams('/user/enterCode', query),
      'GET',
      query
    );
  };

  /**
   * UserService.getLeaveInfo
   *
   * - description:
   * - pathName: /user/leaveInfo
   * - method: GET
   */
  public getLeaveInfo = async (query: IUserService_getLeaveInfo_Query) => {
    return this.request<IUserService_getLeaveInfo_Resp>(
      replaceUrlParams('/user/leaveInfo', query),
      'GET',
      query
    );
  };

  /**
   * UserService.addReserveRequest
   *
   * - description:
   * - pathName: /user/reserve_request
   * - method: POST
   */
  public addReserveRequest = async (query: IUserService_addReserveRequest_Query) => {
    return this.request<IUserService_addReserveRequest_Resp>(
      replaceUrlParams('/user/reserve_request', query),
      'POST',
      query
    );
  };

  /**
   * UserService.cancelReserveRequest
   *
   * - description:
   * - pathName: /user/reserve_request/:id/cancel
   * - method: PUT
   */
  public cancelReserveRequest = async (query: IUserService_cancelReserveRequest_Query) => {
    return this.request<IUserService_cancelReserveRequest_Resp>(
      replaceUrlParams('/user/reserve_request/:id/cancel', query),
      'PUT',
      query
    );
  };
}

export function replaceUrlParams(url: string, data?: any) {
  if (!data) return url;

  const match = url.match(/\\\\\\\\:\\\\\\\\w+/g);
  if (match) {
    const pnSet = new Set(match.map(p => p.replace(/^\\\\\\\\:/, '')));
    pnSet.forEach(p => {
      url = url.replace(new RegExp(\`\\\\:\${p}\`, 'g'), data[p]);
    });
  }

  return url;
}

export const createAllService = (tap?: (s: BaseService) => void) => {
  const result = {
    Page: new PageService(),
    Reserve: new ReserveService(),
    System: new SystemService(),
    User: new UserService(),
  };

  // tap
  Object.values(result).forEach(s => tap?.(s));

  return result;
};
"
`;

exports[`generateAPIService api2: api2 1`] = `
"//banner
import { BaseService } from './BaseService';

export type ISnapshotcontrollerService_create_Query = { device: string; timestamp: string };
export type ISnapshotcontrollerService_create_Resp = string;

export type IFaceindexcontrollerService_getPage_Query = any;
export type IFaceindexcontrollerService_getPage_Resp = {
  pageSize: any;
  pageNum: any;
  total: any;
  list: Array<{
    id: any;
    created_at: string;
    updated_at: string;
    age: any;
    gender: any;
    expression: any;
    snapshot?: {
      id: any;
      created_at: string;
      updated_at: string;
      ossPath: string;
      faceIndexStatus: string;
      device?: { id: any; created_at: string; updated_at: string; name: string; title: string };
    };
  }>;
};

export type IAuthcontrollerService_login_Query = { name: string; password: string };
export type IAuthcontrollerService_login_Resp = {
  accessToken: string;
  user: {
    id: any;
    created_at: string;
    updated_at: string;
    name: string;
    nick: string;
    role: string;
  };
};

export type IUsercontrollerService_getUser_Query = any;
export type IUsercontrollerService_getUser_Resp = {
  id: any;
  created_at: string;
  updated_at: string;
  name: string;
  nick: string;
  role: string;
};

export type IDevicecontrollerService_getPage_Query = any;
export type IDevicecontrollerService_getPage_Resp = {
  pageSize: any;
  pageNum: any;
  total: any;
  list: Array<{ id: any; created_at: string; updated_at: string; name: string; title: string }>;
};

export class SnapshotcontrollerService extends BaseService {
  /**
   * SnapshotcontrollerService.create
   *
   * - description: 快照上传
   * - pathName: /snapshot
   * - method: POST
   */
  public create = async (query: ISnapshotcontrollerService_create_Query) => {
    return this.request<ISnapshotcontrollerService_create_Resp>(
      replaceUrlParams('/snapshot', query),
      'POST',
      query
    );
  };
}

export class FaceindexcontrollerService extends BaseService {
  /**
   * FaceindexcontrollerService.getPage
   *
   * - description:
   * - pathName: /faceIndex
   * - method: GET
   */
  public getPage = async () => {
    return this.request<IFaceindexcontrollerService_getPage_Resp>(
      replaceUrlParams('/faceIndex'),
      'GET'
    );
  };
}

export class AuthcontrollerService extends BaseService {
  /**
   * AuthcontrollerService.login
   *
   * - description:
   * - pathName: /auth/login
   * - method: POST
   */
  public login = async (query: IAuthcontrollerService_login_Query) => {
    return this.request<IAuthcontrollerService_login_Resp>(
      replaceUrlParams('/auth/login', query),
      'POST',
      query
    );
  };
}

export class UsercontrollerService extends BaseService {
  /**
   * UsercontrollerService.getUser
   *
   * - description: 获取用户信息
   * - pathName: /user
   * - method: GET
   */
  public getUser = async () => {
    return this.request<IUsercontrollerService_getUser_Resp>(replaceUrlParams('/user'), 'GET');
  };
}

export class DevicecontrollerService extends BaseService {
  /**
   * DevicecontrollerService.getPage
   *
   * - description:
   * - pathName: /device
   * - method: GET
   */
  public getPage = async () => {
    return this.request<IDevicecontrollerService_getPage_Resp>(replaceUrlParams('/device'), 'GET');
  };
}

export function replaceUrlParams(url: string, data?: any) {
  if (!data) return url;

  const match = url.match(/\\\\\\\\:\\\\\\\\w+/g);
  if (match) {
    const pnSet = new Set(match.map(p => p.replace(/^\\\\\\\\:/, '')));
    pnSet.forEach(p => {
      url = url.replace(new RegExp(\`\\\\:\${p}\`, 'g'), data[p]);
    });
  }

  return url;
}

export const createAllService = (tap?: (s: BaseService) => void) => {
  const result = {
    Snapshotcontroller: new SnapshotcontrollerService(),
    Faceindexcontroller: new FaceindexcontrollerService(),
    Authcontroller: new AuthcontrollerService(),
    Usercontroller: new UsercontrollerService(),
    Devicecontroller: new DevicecontrollerService(),
  };

  // tap
  Object.values(result).forEach(s => tap?.(s));

  return result;
};
"
`;
